runtime/server.go

```diff
@@ -0,0 +1,538 @@
+// Copyright 2016 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package runtime
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/eval"
+	"github.com/open-policy-agent/opa/version"
+	"github.com/pkg/errors"
+)
+
+// apiErrorV1 models an error response sent to the client.
+type apiErrorV1 struct {
+	Code    int
+	Message string
+}
+
+func (err *apiErrorV1) Bytes() []byte {
+	if bs, err := json.MarshalIndent(err, "", "  "); err == nil {
+		return bs
+	}
+	return nil
+}
+
+// patchV1 models a single patch operation against a document.
+type patchV1 struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value"`
+}
+
+// policyV1 models a policy module in OPA.
+type policyV1 struct {
+	ID     string
+	Module *ast.Module
+}
+
+func (p *policyV1) Equal(other *policyV1) bool {
+	return p.ID == other.ID && p.Module.Equal(other.Module)
+}
+
+// resultSetV1 models the result of an ad-hoc query.
+type resultSetV1 []map[string]interface{}
+
+// Server contains runtime state specific to the server-mode persona, e.g.,
+// HTTP router.
+//
+// Notes:
+//
+// - In the future, the HTTP routing could be factored into a separate module
+// relying on the server (which would remain RPC-based). For now, it's simpler
+// to keep the HTTP routing and backend implementation in one place.
+type Server struct {
+	Addr    string
+	Persist bool
+	Runtime *Runtime
+	Router  *mux.Router
+
+	mtx sync.RWMutex
+}
+
+// NewServer returns a new Server.
+func NewServer(rt *Runtime, addr string, persist bool) *Server {
+
+	s := &Server{
+		Addr:    addr,
+		Persist: persist,
+		Runtime: rt,
+		Router:  mux.NewRouter(),
+	}
+
+	s.registerHandlerV1("/data/{path:.+}", "GET", s.v1DataGet)
+	s.registerHandlerV1("/data/{path:.+}", "PATCH", s.v1DataPatch)
+	s.registerHandlerV1("/policies", "GET", s.v1PoliciesList)
+	s.registerHandlerV1("/policies/{id}", "DELETE", s.v1PoliciesDelete)
+	s.registerHandlerV1("/policies/{id}", "GET", s.v1PoliciesGet)
+	s.registerHandlerV1("/policies/{id}/raw", "GET", s.v1PoliciesRawGet)
+	s.registerHandlerV1("/policies/{id}", "PUT", s.v1PoliciesPut)
+	s.registerHandlerV1("/query", "GET", s.v1QueryGet)
+
```

Comment:
Looks like we decided on manually writing the API instead of using a tool like swagger.  Is swagger/etc. an option for the future, or is there something stopping us from using it?